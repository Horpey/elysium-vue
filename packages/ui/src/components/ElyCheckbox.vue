<script lang="ts" setup>
import type { InputHTMLAttributes } from 'vue'

interface Props extends /* @vue-ignore */ InputHTMLAttributes {
  name?: string
  value?: string | number | boolean
  modelValue?: boolean
  required?: boolean
  disabled?: boolean
  inputId?: string
  onClick?: (event: Event) => void
  onChange?: (event: Event) => void
}

const props = withDefaults(defineProps<Props>(), {
  inputId: 'checkbox',
  name: 'checkbox',
  modelValue: undefined,
  required: false,
  disabled: false,
  icon: undefined,
  onClick: () => { },
  onChange: () => { },
})

const emit = defineEmits(['update:modelValue'])

const toggle = computed({
  get() {
    return props.modelValue
  },
  set(value) {
    emit('update:modelValue', value)
  },
})

function onChange(event: Event) {
  props.onChange(event)
  emit('update:modelValue', toggle.value)
}

function onClick(event: Event) {
  props.onClick(event)
}
</script>

<template>
  <div class="relative flex items-start">
    <div class="flex h-6 items-center">
      <input
        v-bind="$attrs"
        :id="inputId"
        v-model="toggle"
        :value="value"
        :name="name"
        type="checkbox"
        class="h-4 w-4 border-gray-300 text-koromiko-700 focus:ring-koromiko-700 disabled:cursor-not-allowed disabled:opacity-60"
        :disabled="disabled"
        :checked="checked"
        :required="required"
        @click="onClick"
        @change="onChange"
      >
    </div>
    <label :for="inputId" class="ml-3 text-sm leading-6">
      <slot />
    </label>
  </div>
</template>
