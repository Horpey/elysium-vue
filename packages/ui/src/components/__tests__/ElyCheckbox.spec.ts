import { mount } from '@vue/test-utils'
import { describe, expect, it, vi } from 'vitest'
import ElyCheckbox from '../ElyCheckbox.vue'

describe('elyCheckbox', () => {
  it('renders with default props', async () => {
    const wrapper = mount(ElyCheckbox)

    // Assert that the component is rendered
    expect(wrapper.exists()).toBe(true)

    // Assert default values
    expect(wrapper.find('input[type="checkbox"]').exists()).toBe(true)
    expect(wrapper.find('input[type="checkbox"]').attributes('id')).toBe('checkbox')
    expect(wrapper.find('input[type="checkbox"]').attributes('name')).toBe('checkbox')
    expect(wrapper.find('input[type="checkbox"]').attributes('disabled')).toBe(undefined)
    expect(wrapper.find('input[type="checkbox"]').classes()).toContain('size-4')
    // ... add more assertions based on your styling and default props

    // Assert slot content is not provided
    expect(wrapper.text()).toBe('')

    // Assert snapshot
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders with custom props and triggers click event', async () => {
    const wrapper = mount(ElyCheckbox, {
      props: {
        inputId: 'custom-checkbox',
        name: 'custom-checkbox',
        modelValue: true,
        required: true,
        disabled: false,
        value: 'custom-value',
        onClick: vi.fn(),
        onChange: vi.fn(),
      },
      slots: {
        default: 'Custom Checkbox Label',
      },
    })

    // Assert that the component is rendered
    expect(wrapper.exists()).toBe(true)

    // Assert custom values
    expect(wrapper.find('input[type="checkbox"]').attributes('id')).toBe('custom-checkbox')
    expect(wrapper.find('input[type="checkbox"]').attributes('name')).toBe('custom-checkbox')
    expect(wrapper.find('input[type="checkbox"]').attributes('value')).toBe('custom-value')
    expect(wrapper.find('input[type="checkbox"]').attributes('disabled')).toBe(undefined)

    // Trigger click event
    await wrapper.find('input[type="checkbox"]').trigger('click')

    // Assert that the onClick function was called
    expect(wrapper.props().onClick).toHaveBeenCalled()

    // Assert snapshot
    expect(wrapper.html()).toMatchSnapshot()
  })

  // Add more test cases as needed
})
