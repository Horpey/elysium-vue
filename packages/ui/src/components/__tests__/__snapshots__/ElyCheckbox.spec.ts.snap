// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`elyCheckbox > renders with custom props and triggers click event 1`] = `
"<div class="relative flex items-start">
  <div class="flex h-6 items-center"><input id="custom-checkbox" name="custom-checkbox" type="checkbox" class="h-4 w-4 border-gray-300 text-koromiko-700 focus:ring-koromiko-700 disabled:cursor-not-allowed disabled:opacity-60" required="" value="custom-value"></div><label for="custom-checkbox" class="ml-3 text-sm leading-6">Custom Checkbox Label</label>
</div>"
`;

exports[`elyCheckbox > renders with default props 1`] = `
"<div class="relative flex items-start">
  <div class="flex h-6 items-center"><input id="checkbox" name="checkbox" type="checkbox" class="h-4 w-4 border-gray-300 text-koromiko-700 focus:ring-koromiko-700 disabled:cursor-not-allowed disabled:opacity-60" value="false"></div><label for="checkbox" class="ml-3 text-sm leading-6"></label>
</div>"
`;
